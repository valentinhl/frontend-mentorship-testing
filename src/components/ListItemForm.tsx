import { FC } from 'react';
import { ListItemType } from './models';
import { v4 as uuidv4 } from 'uuid';
import {
  descriptionLabel,
  submitButtonText,
  titleLabel
} from './ListItemUIText';

export type ListFormOnSubmit = (newValue: ListItemType) => void;

export interface ListItemFormProps {
  onSubmit: ListFormOnSubmit;
  isSubmitDisabled?: boolean;
}

export interface ListItemFormData {
  title: { value: string };
  description: { value: string };
}

const ListItemForm: FC<ListItemFormProps> = ({
  onSubmit,
  isSubmitDisabled = false
}): JSX.Element => {
  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const { title, description } = (
      event.target as EventTarget & HTMLFormElement
    ).elements as HTMLFormControlsCollection & ListItemFormData;

    onSubmit({
      id: uuidv4(),
      title: title.value,
      description: description.value
    });
  };

  return (
    <form onSubmit={event => submitForm(event)} aria-label='form'>
      <br /> <br />
      <label htmlFor='title'>{titleLabel}</label>
      <input id='title' name='title' />
      <br /> <br />
      <label htmlFor='description'>{descriptionLabel}</label>
      <textarea id='description' name='description' />
      <br /> <br />
      <button disabled={isSubmitDisabled}>{submitButtonText}</button>
    </form>
  );
};

export default ListItemForm;
