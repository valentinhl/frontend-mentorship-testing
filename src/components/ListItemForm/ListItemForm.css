.todo-creation-form {
  display: flex;
  flex-direction: column;
  background-color: rgba(255, 255, 255, 0.75);
  border-radius: 32px;
  padding: 32px;
  margin-top: 32px;
  min-width: 350px;
}

.fieldset {
  border: none;
  display: flex;
  justify-content: space-between;
}

.input-field {
  width: 200px;
}

.error {
  display: 'block';
  color: red;
}
