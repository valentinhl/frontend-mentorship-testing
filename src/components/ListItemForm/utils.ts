import { CreateEmptyErrorList, ListItemFormErrorList } from '../models';

export const isRequired = (
  errors: ListItemFormErrorList,
  field: 'title' | 'description',
  value: string
): ListItemFormErrorList => {
  const allowedFields = ['title', 'description'];
  const newErrors = { ...errors };
  if (allowedFields.includes(field)) {
    const fieldError = !value.length ? `${field} must be entered` : '';
    newErrors[field] = fieldError;
  }
  return newErrors;
};

export const createEmptyErrorList: CreateEmptyErrorList = () => ({
  title: '',
  description: ''
});
