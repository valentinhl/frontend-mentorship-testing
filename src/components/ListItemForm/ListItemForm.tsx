import { FC, useState } from 'react';
import {
  ListFormOnChange,
  ListFormOnSubmit,
  ListItemFormData,
  ListItemFormErrorList,
  ListItemType
} from '../models';
import { v4 as uuidv4 } from 'uuid';
import {
  addListItemForm,
  descriptionLabel,
  submitButtonText,
  titleLabel
} from '../ListItemUIText';
import './ListItemForm.css';
import { createEmptyErrorList, isRequired } from './utils';
export interface ListItemFormProps {
  formItem: ListItemType;
  onChange: ListFormOnChange;
  onSubmit: ListFormOnSubmit;
  isSubmitDisabled?: boolean;
}

const ListItemForm: FC<ListItemFormProps> = ({
  formItem,
  onChange,
  onSubmit,
  isSubmitDisabled = false
}): JSX.Element => {
  const [errors, setErrors] = useState<ListItemFormErrorList>(
    createEmptyErrorList()
  );

  const validate = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    name: string,
    value: string
  ) => {
    setErrors(isRequired(errors, name as 'title' | 'description', value));
  };

  const handleChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    event.persist();

    const name = event.target.name;
    const val = event.target.value;

    validate(event, name, val);

    const newValue = { ...formItem };
    newValue[name] = val;

    onChange(newValue);
  };

  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const { title, description } = (
      event.target as EventTarget & HTMLFormElement
    ).elements as HTMLFormControlsCollection & ListItemFormData;

    onSubmit({
      id: formItem.id ? formItem.id : uuidv4(),
      title: title.value,
      description: description.value
    });
  };

  return (
    <form
      name={addListItemForm}
      className='todo-creation-form'
      onSubmit={event => submitForm(event)}
    >
      <fieldset className='fieldset'>
        <label htmlFor='title'>{titleLabel}</label>
        <input
          className='input-field'
          id='title'
          name='title'
          onChange={handleChange}
          value={formItem.title}
        />
        {errors.title && <span className='error'>{errors.title}</span>}
      </fieldset>

      <fieldset className='fieldset'>
        <label htmlFor='description'>{descriptionLabel}</label>
        <textarea
          className='input-field'
          id='description'
          placeholder='Please enter a description here'
          name='description'
          onChange={handleChange}
          value={formItem.description}
        />
        {errors.description && (
          <span className='error'>{errors.description}</span>
        )}
      </fieldset>

      <button disabled={isSubmitDisabled}>{submitButtonText}</button>
    </form>
  );
};

export default ListItemForm;
