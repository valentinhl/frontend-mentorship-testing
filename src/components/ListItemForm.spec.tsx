import { render, screen, cleanup, RenderResult } from '@testing-library/react';
import ListItemForm, { ListFormOnSubmit } from './ListItemForm';
import { MOCK_STRING } from '../testing/mockValues';
import { changeInputValue, submitForm } from '../testing/utils';
import {
  descriptionLabel,
  submitButtonText,
  titleLabel
} from './ListItemUIText';

const getMockId = () => 'id';

jest.mock('uuid', () => ({
  ...jest.requireActual('uuid'),
  v4: getMockId
}));

describe('ListItemForm.tsx', () => {
  const MOCK_ID = getMockId();
  const onSubmitMock = jest.fn() as jest.MockedFunction<ListFormOnSubmit>;
  let container: HTMLElement;
  let rerenderComponent: RenderResult['rerender'];

  beforeEach(() => {
    const { rerender } = render(<ListItemForm onSubmit={onSubmitMock} />);
    rerenderComponent = rerender;
  });

  afterEach(() => {
    onSubmitMock.mockClear();
    cleanup();
  });

  it('renders the component successfully', () => {
    expect(screen.getByRole('form')).toMatchSnapshot();
  });

  it('submit is enabled by default', () => {
    expect(screen.getByText(submitButtonText)).toBeEnabled();
  });

  it('submit is enabled when isSubmitDisabled is passed', () => {
    rerenderComponent(
      <ListItemForm onSubmit={onSubmitMock} isSubmitDisabled />
    );
    expect(screen.getByText(submitButtonText)).toBeDisabled();
  });

  it('submit is enabled when isSubmitDisabled is false', () => {
    rerenderComponent(
      <ListItemForm onSubmit={onSubmitMock} isSubmitDisabled={false} />
    );
    expect(screen.getByText(submitButtonText)).toBeEnabled();
  });

  it('calls onSubmit prop once when submit event fires', () => {
    submitForm(screen, submitButtonText);
    expect(onSubmitMock).toHaveBeenCalledTimes(1);
  });

  it('onSubmit receives object with right values', () => {
    changeInputValue(screen, titleLabel, MOCK_STRING);
    changeInputValue(screen, descriptionLabel, MOCK_STRING);
    submitForm(screen, submitButtonText);

    expect(onSubmitMock).toHaveBeenCalledWith({
      id: MOCK_ID,
      title: MOCK_STRING,
      description: MOCK_STRING
    });
  });
});
