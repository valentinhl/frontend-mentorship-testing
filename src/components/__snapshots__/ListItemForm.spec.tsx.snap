// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemForm.tsx renders the component successfully 1`] = `
<form
  aria-label="form"
>
  <br />
   
  <br />
  <label
    for="title"
  >
    Title:
  </label>
  <input
    id="title"
    name="title"
  />
  <br />
   
  <br />
  <label
    for="description"
  >
    Description:
  </label>
  <textarea
    id="description"
    name="description"
  />
  <br />
   
  <br />
  <button>
    SUBMIT
  </button>
</form>
`;
